pkg/pushsync/pushsync.go

```diff
@@ -77,59 +78,139 @@ func (ps *PushSync) Close() error {
 	return nil
 }
 
-// handler handles chunk delivery from other node and inserts it to localstore.
-// it also sends this chunk to the closest peer if one exists.
+// handler handles chunk delivery from other node and forwards to its destination node.
+// If the current node is the destination, it stores in the local store and sends a receipt.
 func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
-	_, r := protobuf.NewWriterAndReader(stream)
+	w, r := protobuf.NewWriterAndReader(stream)
 	defer stream.Close()
 
-	var ch pb.Delivery
+	// Get the delivery
+	ch, err := ps.getChunkDelivery(r)
+	if err != nil {
+		return fmt.Errorf("could not receive chunk delivery: %w ", err)
+	}
 
-	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
+	// create chunk
+	addr := swarm.NewAddress(ch.Address)
+	chunk := swarm.NewChunk(addr, ch.Data)
+
+	// Select the closest peer to forward the chunk
+	peer, err := ps.peerSuggester.ClosestPeer(chunk.Address())
+	if err != nil {
+		// If i am the closest peer then store the chunk and send receipt
+		if errors.Is(err, topology.ErrWantSelf) {
+
+			// Store the chunk in the local store
+			_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+			if err != nil {
+				return fmt.Errorf("could not store chunk in local DB: %w ", err)
+			}
+			ps.metrics.TotalChunksStoredInDB.Inc()
+
+			// Send a receipt immediately once the storage of the chunk is successfull
+			err = ps.sendReceipt(w, chunk.Address())
+			if err != nil {
+				return fmt.Errorf("could not send receipt: %w ", err)
+			}
 			return nil
 		}
 		return err
 	}
 
-	// create chunk and store it in the local store
-	addr := swarm.NewAddress(ch.Address)
-	chunk := swarm.NewChunk(addr, ch.Data)
-	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
-	if err != nil {
+	//Dont forward this chunk to the same peer as the received peer
+	if p.Address.Equal(peer) {
 		return err
 	}
 
-	// push this to the closest node too
-	peer, err := ps.peerSuggester.ClosestPeer(addr)
+	// Forward chunk to closest peer
+	streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
 	if err != nil {
-		if errors.Is(err, topology.ErrWantSelf) {
-			// i'm the closest - nothing to do
-			return nil
-		}
+		return fmt.Errorf("new stream: %w", err)
+	}
+	defer streamer.Close()
+
+	wc, rc := protobuf.NewWriterAndReader(streamer)
+
+	if err := ps.sendChunkDelivery(wc, chunk); err != nil {
+		return fmt.Errorf("could not send chunk to closest peer: %w ", err)
```

Comment:
The error annotation is better without `could not` in it as errors are chained.

Also, there is a blank space at the end of the message that should be removed. I believe that this is a copy/paste error from the chat, that was corrected in the chat.